### ISSUE TITLE
How to realize the loss curves of badam and lora in the paper

### ISSUE THREAD

**Issue by lovekdl** [Ledzy/BAdam#5]
Hello, I am very interested in your work and tried to reproduce the experiments in the paper.
I followed the the experiment setup according to the section 3.1 in the paper. The learning rate is 1e-5, the batch size is 8, the gradient accumulation is 15, and the lora rank is 100.

<details> 
    <summary>Here is my badam config </summary>

```
#!/bin/bash
CUDA_VISIBLE_DEVICES=0 python ../../src/train_bash.py \
    --stage sft \
    --do_train \
    --model_name_or_path meta-llama/Llama-2-7b-hf \
    --dataset alpaca_gpt4_en \
    --dataset_dir ../../data \
    --template default \
    --finetuning_type full \
    --use_badam \
    --badam_switch_mode ascending \
    --badam_switch_block_every 10 \
    --badam_verbose 2 \
    --output_dir ../../saves/LLaMA2-7B/badam/sft \
    --overwrite_cache \
    --overwrite_output_dir \
    --cutoff_len 1024 \
    --preprocessing_num_workers 16 \
    --per_device_train_batch_size 8 \
    --gradient_accumulation_steps 15 \
    --lr_scheduler_type cosine \
    --logging_steps 1 \
    --warmup_steps 0 \
    --save_steps 1000 \
    --learning_rate 1e-5 \
    --report_to wandb \
    --num_train_epochs 7 \
    --plot_loss \
    --bf16

```
</details>

<details> 
    <summary>Here is my lora config </summary>

```
#!/bin/bash
CUDA_VISIBLE_DEVICES=0 python ../../src/train_bash.py \
    --stage sft \
    --do_train \
    --model_name_or_path meta-llama/Llama-2-7b-hf \
    --dataset alpaca_gpt4_en \
    --dataset_dir ../../data \
    --template default \
    --finetuning_type lora \
    --lora_rank 100 \
    --output_dir ../../saves/LLaMA2-7B/lora/lora-llama-2-7b-r16-batch8-acc15-epoch7-1e-5 \
    --overwrite_cache \
    --overwrite_output_dir \
    --cutoff_len 1024 \
    --preprocessing_num_workers 16 \
    --per_device_train_batch_size 8 \
    --gradient_accumulation_steps 15 \
    --lr_scheduler_type cosine \
    --logging_steps 1 \
    --warmup_steps 0 \
    --save_steps 1000 \
    --report_to wandb \
    --learning_rate 1e-5 \
    --num_train_epochs 7 \
    --plot_loss \
    --bf16


```
</details>

When I ran the experiment, I found that badam method cost about 23G memory, while lora method cost about 39G memory on A100-80G (but only cost 24G on RTX3090-24G). I am wondering if my experiment config is correct.

I expected the training loss curves like this:
![image](https://github.com/Ledzy/BAdam/assets/44801395/e985e3ef-37ad-4079-b657-8c1c7ab2197a)
But what I actually got was this( purple for badam and blue for lora):
![image](https://github.com/Ledzy/BAdam/assets/44801395/3af963cf-b012-4b77-a3d6-12cf76590ac0)

**Comment by Ledzy** [comment]
Thanks for your interest and your efforts in reproducing our experimental result. 

In our paper, BAdam's convergence curve is based on $K=50$, i.e. `badam_switch_block_every`=50. This hyperparameter determines the number of Adam update steps for each block. In order to fully utilize the advantage of Adam's acceleration when solving each block sub-problem, we suggest to set it to be above 50; see our [hyperparameter suggestion](https://github.com/Ledzy/BAdam?tab=readme-ov-file#hyperparameter-suggestion) for more discussion.

As for memory issue, you may check the `torch.cuda.memory_allocated()` for the actual memory occupied by tensors (instead of the value returned by `nvidia-smi`), which should align with our reported memory consumption. The memory gap observed on A100 may due to that torch reserved memory (return by `torch.cuda.memory_reserved`) is much larger than the allocated memory; use `torch.cuda.empty_cache()` to empty that part, if necessary.

**Comment by lovekdl** [comment]
@Ledzy Thanks for your reply! I have tried setting the `badam_switch_block_every`=50, and I got the loss curves :
![image](https://github.com/Ledzy/BAdam/assets/44801395/4c932104-2a71-4f91-859a-fa1810913134)
The performance of BAdam improved but still worse than LoRA.
For memory issue, I used `torch.cuda.max_memory_allocated()` to measure memory consumtion as you suggested, the memory consumption of BAdam is 19.09GB, the memory consumption of LoRA is 20.87GB.
I would appreciate it if you could provide a detailed script of BAdam and LoRA matching the loss curves in the paper. (●'◡'●)

**Comment by Ledzy** [comment]
Please use the following command to run our uploaded code **under folder "llama-alpaca"**. It should produce exactly the same loss curve shown in Figure 2 of our paper (version 1, uploaded on 2024.4.3):

```bash
CUDA_VISIBLE_DEVICES=0 python src/train_bash.py \
    --stage sft \
    --model_name_or_path your_llama2_model_path \
    --do_train \
    --dataset alpaca_gpt4_en \
    --template default \
    --finetuning_type block \
    --output_dir ./outputs/BAdam_1e-5_full \
    --overwrite_cache \
    --per_device_train_batch_size 8 \
    --per_device_eval_batch_size 5 \
    --gradient_accumulation_steps 15 \
    --lr_scheduler_type cosine \
    --logging_steps 1 \
    --save_steps 10000 \
    --val_size 500 \
    --eval_steps 100 \
    --evaluation_strategy steps \
    --learning_rate 1e-5 \
    --num_train_epochs 32 \
    --overwrite_output_dir \
    --plot_loss \
    --switch_block_every 50 \
    --switch_mode ascending \
    --bf16
```
The reason why BAdam is slow in your experiment may attribute to fast decay of learning rate, which goes to 0 in epoch 7. Try to set "num_train_epochs" to 32, or use constant learning rate. Note that lower training loss does not necessarily indicates better downstream performance, e.g. MT-bench score.
